Thanks, will test there later. While compiling on another PC, I see a warning:

[158/255 2.3/sec] Building CXX object src\CMakeFiles\stelMain.dir\core\modules\AtmosphereLightweight.cpp.obj
D:\StelDEV\stellarium\src\core\modules\AtmosphereLightweight.cpp(657): warning C4312: 'reinterpret_cast': conversion from 'const _Ty' to 'void *' of greater size
        with
        [
            _Ty=std::seed_seq::result_type
        ]

Is that clean?

--- a/src/core/TextureAverageComputer.cpp
+++ b/src/core/TextureAverageComputer.cpp
@@ -20,6 +20,7 @@
 #include "TextureAverageComputer.hpp"
 #include "StelMainView.hpp"
 #include "StelUtils.hpp"
+#include <array>
 
 namespace
 {
@@ -64,7 +65,16 @@ Vec4f TextureAverageComputer::getCurrentTextureDeepestMipLevelPixelGLES(const in
 
 	blitTexProgram->release();
 	Vec4f pixel(NAN,NAN,NAN,NAN);
-	GL(gl.glReadPixels(0,0,1,1,GL_RGBA,GL_FLOAT,&pixel[0]));
+	if (textureIsFloat)
+	{
+		GL(gl.glReadPixels(0,0,1,1,GL_RGBA,GL_FLOAT,&pixel[0]));
+	}
+	else
+	{
+		std::array<uint8_t, 4> data{};
+		GL(gl.glReadPixels(0,0,1,1,GL_RGBA,GL_UNSIGNED_BYTE,&data[0]));
+		pixel = {data[0] / 255.f, data[1] / 255.f, data[2] / 255.f, data[3] / 255.f};
+	}
 
 	GL(gl.glViewport(oldViewport[0], oldViewport[1], oldViewport[2], oldViewport[3]));
 	GL(gl.glBindFramebuffer(GL_FRAMEBUFFER, oldFBO));
@@ -215,10 +225,11 @@ void TextureAverageComputer::checkNeedForWorkaround()
 }
 
 // Clobbers: GL_TEXTURE_BINDING_2D, GL_VERTEX_ARRAY_BINDING, GL_ARRAY_BUFFER_BINDING
-TextureAverageComputer::TextureAverageComputer(StelOpenGL::HighGraphicsFunctions& gl, const int texWidth, const int texHeight, const GLenum internalFormat)
+TextureAverageComputer::TextureAverageComputer(StelOpenGL::HighGraphicsFunctions& gl, const int texWidth, const int texHeight, const GLenum internalFormat, const bool textureIsFloat)
 	: gl(gl)
 	, npotWidth(texWidth)
 	, npotHeight(texHeight)
+	, textureIsFloat(textureIsFloat)
 {
 	isGLES = StelMainView::getInstance().getGLInformation().isGLES;
 
@@ -287,7 +298,8 @@ void main()
 		GL(gl.glBindTexture(GL_TEXTURE_2D, potTex));
 		const auto potWidth  = StelUtils::getSmallerPowerOfTwo(npotWidth);
 		const auto potHeight = StelUtils::getSmallerPowerOfTwo(npotHeight);
-		GL(gl.glTexImage2D(GL_TEXTURE_2D,0,internalFormat,potWidth,potHeight,0,GL_RGBA,GL_FLOAT,nullptr));
+		GL(gl.glTexImage2D(GL_TEXTURE_2D,0,internalFormat,potWidth,potHeight,0,GL_RGBA,
+		                   textureIsFloat ? GL_FLOAT : GL_UNSIGNED_BYTE,nullptr));
 		GL(gl.glBindTexture(GL_TEXTURE_2D,0));
 		GL(gl.glBindFramebuffer(GL_DRAW_FRAMEBUFFER,potFBO));
 		GL(gl.glFramebufferTexture2D(GL_DRAW_FRAMEBUFFER,GL_COLOR_ATTACHMENT0,GL_TEXTURE_2D,potTex,0));

--- a/src/core/TextureAverageComputer.cpp
+++ b/src/core/TextureAverageComputer.cpp
@@ -21,15 +21,16 @@
 #include "StelMainView.hpp"
 #include "StelUtils.hpp"
 #include <array>
+#include <QOpenGLExtraFunctions>
 
 namespace
 {
 class FBORestorer
 {
-	StelOpenGL::HighGraphicsFunctions& gl;
+	QOpenGLExtraFunctions& gl;
 	GLint oldFBO=-1;
 public:
-	FBORestorer(StelOpenGL::HighGraphicsFunctions& gl)
+	FBORestorer(QOpenGLExtraFunctions& gl)
 		: gl(gl)
 	{
 	}
@@ -100,7 +101,7 @@ Vec4f TextureAverageComputer::getCurrentTextureDeepestMipLevelPixelGL(const int 
 #endif
 
 	Vec4f pixel;
-	GL(gl.glGetTexImage(GL_TEXTURE_2D, deepestLevel, GL_RGBA, GL_FLOAT, &pixel[0]));
+	GL(hiGL->glGetTexImage(GL_TEXTURE_2D, deepestLevel, GL_RGBA, GL_FLOAT, &pixel[0]));
 	return pixel;
 #else
 	return {NAN, NAN, NAN, NAN};
@@ -225,13 +226,14 @@ void TextureAverageComputer::checkNeedForWorkaround()
 }
 
 // Clobbers: GL_TEXTURE_BINDING_2D, GL_VERTEX_ARRAY_BINDING, GL_ARRAY_BUFFER_BINDING
-TextureAverageComputer::TextureAverageComputer(StelOpenGL::HighGraphicsFunctions& gl, const int texWidth, const int texHeight, const GLenum internalFormat, const bool textureIsFloat)
-	: gl(gl)
+TextureAverageComputer::TextureAverageComputer(StelOpenGL::HighGraphicsFunctions* hiGL, const int texWidth, const int texHeight, const GLenum internalFormat, const bool textureIsFloat)
+	: gl(*QOpenGLContext::currentContext()->extraFunctions())
+	, hiGL(hiGL)
 	, npotWidth(texWidth)
 	, npotHeight(texHeight)
 	, textureIsFloat(textureIsFloat)
 {
-	isGLES = StelMainView::getInstance().getGLInformation().isGLES;
+	isGLES = StelMainView::getInstance().getGLInformation().isGLES || !hiGL;
 
 	GL(gl.glGenVertexArrays(1, &vao));
 	GL(gl.glBindVertexArray(vao));

--- a/src/core/TextureAverageComputer.hpp
+++ b/src/core/TextureAverageComputer.hpp
@@ -38,6 +38,7 @@ class TextureAverageComputer
 	GLuint potTex = 0;
 	GLuint vbo = 0, vao = 0;
 	GLint npotWidth, npotHeight;
+	bool textureIsFloat = false;
 	bool isGLES = false;
 	static inline bool needForWorkaroundChecked = false;
 	static inline bool npotWorkaroundNeeded = false;
@@ -51,7 +52,8 @@ public:
 	Vec4f getTextureAverage(GLuint texture);
 	//!< Can be used for power-of-two textures
 	Vec4f getTextureAverageSimple(GLuint texture, int width, int height);
-	TextureAverageComputer(StelOpenGL::HighGraphicsFunctions&, int texW, int texH, GLenum internalFormat);
+	TextureAverageComputer(StelOpenGL::HighGraphicsFunctions&, int texW, int texH,
+	                       GLenum internalFormat, bool textureIsFloat);
 	~TextureAverageComputer();
 };
 

--- a/src/core/TextureAverageComputer.hpp
+++ b/src/core/TextureAverageComputer.hpp
@@ -30,7 +30,8 @@
 
 class TextureAverageComputer
 {
-	StelOpenGL::HighGraphicsFunctions& gl;
+	QOpenGLExtraFunctions& gl;
+	StelOpenGL::HighGraphicsFunctions* hiGL;
 	std::unique_ptr<QOpenGLShaderProgram> blitTexProgram;
 	//! This FBO is used for getting the deepest mipmap levels of textures in GLES mode.
 	std::unique_ptr<QOpenGLFramebufferObject> glesFBO;
@@ -52,7 +53,7 @@ public:
 	Vec4f getTextureAverage(GLuint texture);
 	//!< Can be used for power-of-two textures
 	Vec4f getTextureAverageSimple(GLuint texture, int width, int height);
-	TextureAverageComputer(StelOpenGL::HighGraphicsFunctions&, int texW, int texH,
+	TextureAverageComputer(StelOpenGL::HighGraphicsFunctions*, int texW, int texH,
 	                       GLenum internalFormat, bool textureIsFloat);
 	~TextureAverageComputer();
 };

--- a/src/core/modules/AtmosphereLightweight.cpp
+++ b/src/core/modules/AtmosphereLightweight.cpp
@@ -93,7 +93,7 @@ AtmosphereLightweight::AtmosphereLightweight()
 	, luminanceProbeFBO_(new QOpenGLFramebufferObject(LUM_PROBE_FBO_WIDTH, LUM_PROBE_FBO_HEIGHT,
 	                                                  QOpenGLFramebufferObject::NoAttachment,
 	                                                  GL_TEXTURE_2D, GL_RGBA))
-	, textureAverager_(new TextureAverageComputer(*StelOpenGL::highGraphicsFunctions(),
+	, textureAverager_(new TextureAverageComputer(StelOpenGL::highGraphicsFunctions(),
 	                                              LUM_PROBE_FBO_WIDTH, LUM_PROBE_FBO_HEIGHT,
 	                                              GL_RGBA, false))
 {

--- a/src/core/modules/AtmosphereLightweight.cpp
+++ b/src/core/modules/AtmosphereLightweight.cpp
@@ -28,6 +28,7 @@
 #include "StelToneReproducer.hpp"
 #include "TextureAverageComputer.hpp"
 
+#include <limits>
 #include <cstring>
 #include <stdexcept>
 #include <QDir>
@@ -121,10 +122,11 @@ AtmosphereLightweight::AtmosphereLightweight()
 		"ATTRIBUTE mediump float color_Y;\n"
 		"ATTRIBUTE mediump vec2 azimuthAndElevation;\n"
 		"VARYING mediump vec3 skyColor;\n"
+		"uniform mediump float colorScale;\n"
 		"void main()\n"
 		"{\n"
 		"	gl_Position = vec4(2. * azimuthAndElevation - 1., 0., 1.);\n"
-		"	skyColor=vec3(color_xy.x, color_xy.y, color_Y);\n"
+		"	skyColor=vec3(color_xy.x, color_xy.y, color_Y * colorScale);\n"
 		"}\n";
 	if (!vShader.compileSourceCode(vSrc))
 		qFatal("Error while compiling atmosphere vertex shader: %s", vShader.log().toUtf8().constData());
@@ -151,6 +153,8 @@ AtmosphereLightweight::AtmosphereLightweight()
 	preparationShaderProgram_->bindAttributeLocation("color_xy", COLOR_XY_ATTRIB_LOC);
 	preparationShaderProgram_->bindAttributeLocation("color_Y", COLOR_Y_ATTRIB_LOC);
 	StelPainter::linkProg(preparationShaderProgram_.get(), "Lightweight atmosphere");
+
+	GL(prepProgramUniformLocations_.colorScale = preparationShaderProgram_->uniformLocation("colorScale"));
 
 	preparationVBO_.setUsagePattern(QOpenGLBuffer::StaticDraw);
 	preparationVBO_.create();
@@ -210,6 +214,12 @@ void AtmosphereLightweight::loadMesh()
 			allVertices.resize(allVertices.size() + numVertices);
 			read(file, path, allVertices.data() + oldVertSize, numVertices * sizeof allVertices[0]);
 			layerVertexOffsetsInVBO_.push_back(oldVertSize * sizeof allVertices[0]);
+			const auto max = std::max_element(allVertices.data() + oldVertSize,
+			                                  allVertices.data() + oldVertSize + numVertices,
+			                                  [](const auto& a, const auto& b)
+			                                  { return a.color_Y < b.color_Y; });
+			constexpr float maxPossibleColorY = std::numeric_limits<decltype(max->color_Y)>::max();
+			layerValueMaxima_.push_back(max->color_Y / maxPossibleColorY);
 
 			const auto oldIndSize = allIndices.size();
 			allIndices.resize(allIndices.size() + numIndices);
@@ -239,6 +249,7 @@ void AtmosphereLightweight::loadMesh()
 
 		vaos_.clear();
 		layerNorms_.clear();
+		layerValueMaxima_.clear();
 		layerSolarElevations_.clear();
 		layerVertexOffsetsInVBO_.clear();
 		layerIndexOffsetsInBuffer_.clear();
@@ -523,7 +534,11 @@ float AtmosphereLightweight::computeAverageLuminance(const StelCore* core)
 	for (int i = 0; i < DRAW_PARAM_COUNT; ++i)
 	{
 		const auto& p = drawParams[i];
-		lum += texAvg[i] * p.relativeBrightness * p.eclipseFactor * p.fboColorScale;
+		auto currLum = texAvg[i] * p.relativeBrightness * p.eclipseFactor * p.fboColorScale;
+		const float maxValue = std::max(layerValueMaxima_[p.layerToDrawA],
+		                                layerValueMaxima_[p.layerToDrawB]);
+		currLum *= maxValue; // restore the scale we used when creating the preparation FBO
+		lum += currLum;
 	}
 	return lum;
 }
@@ -646,6 +661,8 @@ void AtmosphereLightweight::computeColor(StelCore* core, const double JD, const 
 	{
 		const auto& p = drawParams[i];
 		p.fbo->bind();
+		const float maxValue = std::max(layerValueMaxima_[p.layerToDrawA],
+		                                layerValueMaxima_[p.layerToDrawB]);
 		for (const auto layer : {p.layerToDrawA, p.layerToDrawB})
 		{
 			if (layer == p.layerToDrawB)
@@ -658,6 +675,8 @@ void AtmosphereLightweight::computeColor(StelCore* core, const double JD, const 
 			{
 				GL(gl.glDisable(GL_BLEND));
 			}
+			preparationShaderProgram_->setUniformValue(prepProgramUniformLocations_.colorScale,
+			                                           1.f / maxValue);
 			bindVAO(layer);
 			GL(indexBuffer_.bind());
 			const auto indicesOffset = layerIndexOffsetsInBuffer_[layer];
@@ -739,10 +758,15 @@ void AtmosphereLightweight::draw(StelCore*const core)
 	for (int i = 0; i < DRAW_PARAM_COUNT; ++i)
 	{
 		const auto& p = drawParams[i];
+		const float maxValue = std::max(layerValueMaxima_[p.layerToDrawA],
+		                                layerValueMaxima_[p.layerToDrawB]);
 		GL(gl.glBindTexture(GL_TEXTURE_2D, p.fbo->texture()));
-		renderShaderProgram_->setUniformValue(renderUniformLocations_.colorScale,
-		                                      p.fboColorScale * p.relativeBrightness * p.eclipseFactor);
 		renderShaderProgram_->setUniformValue(renderUniformLocations_.sunDir, p.dir);
+
+		float colorScale = p.fboColorScale * p.relativeBrightness * p.eclipseFactor;
+		colorScale *= maxValue; // restore the scale we used when creating the preparation FBO
+		renderShaderProgram_->setUniformValue(renderUniformLocations_.colorScale, colorScale);
+
 		GL(gl.glDrawArrays(GL_TRIANGLE_STRIP, 0, 4));
 	}
 

--- a/src/core/modules/AtmosphereLightweight.cpp
+++ b/src/core/modules/AtmosphereLightweight.cpp
@@ -92,10 +92,10 @@ AtmosphereLightweight::AtmosphereLightweight()
 	               QOpenGLFramebufferObject::NoAttachment, GL_TEXTURE_2D, getInternalFormatForFBO()))
 	, luminanceProbeFBO_(new QOpenGLFramebufferObject(LUM_PROBE_FBO_WIDTH, LUM_PROBE_FBO_HEIGHT,
 	                                                  QOpenGLFramebufferObject::NoAttachment,
-	                                                  GL_TEXTURE_2D, getInternalFormatForFBO()))
+	                                                  GL_TEXTURE_2D, GL_RGBA))
 	, textureAverager_(new TextureAverageComputer(*StelOpenGL::highGraphicsFunctions(),
 	                                              LUM_PROBE_FBO_WIDTH, LUM_PROBE_FBO_HEIGHT,
-	                                              getInternalFormatForFBO()))
+	                                              GL_RGBA, false))
 {
 	setFadeDuration(1.5f);
 
@@ -404,7 +404,7 @@ void main()
 VARYING mediump vec3 ndcPos;
 uniform mat4 projectionMatrixInverse;
 uniform vec3 sunDir;
-uniform highp float colorScale;
+uniform bool isMoon;
 uniform sampler2D atmoTex;
 const highp float PI = 3.14159265;
 
@@ -430,7 +430,7 @@ void main()
 	elevTC = sqrt(elevTC); // elevation coordinate is stretched by squaring it
 	float azimuthTC = relAzimuth / PI;
 	vec3 color_xyY = texture2D(atmoTex, vec2(azimuthTC, elevTC)).xyz;
-	FRAG_COLOR = vec4(color_xyY.xy, color_xyY.z * colorScale, 1);
+	FRAG_COLOR = vec4(isMoon ? 0. : color_xyY.z, isMoon ? color_xyY.z : 0., 0, 1);
 }
 )";
 
@@ -453,9 +453,9 @@ void main()
 	luminanceProbeProgram_->bindAttributeLocation("screenVertex", SCREEN_VERTEX_ATTRIB_LOC);
 	StelPainter::linkProg(luminanceProbeProgram_.get(), "Lightweight atmosphere luminance probe");
 
+	GL(luminanceProbeUniformLocations_.isMoon                  = luminanceProbeProgram_->uniformLocation("isMoon"));
 	GL(luminanceProbeUniformLocations_.sunDir                  = luminanceProbeProgram_->uniformLocation("sunDir"));
 	GL(luminanceProbeUniformLocations_.atmoTex                 = luminanceProbeProgram_->uniformLocation("atmoTex"));
-	GL(luminanceProbeUniformLocations_.colorScale              = luminanceProbeProgram_->uniformLocation("colorScale"));
 	GL(luminanceProbeUniformLocations_.projectionMatrixInverse = luminanceProbeProgram_->uniformLocation("projectionMatrixInverse"));
 
 	bindVAO(VAO_RENDER);
@@ -497,8 +497,8 @@ float AtmosphereLightweight::computeAverageLuminance(const StelCore* core)
 	for (int i = 0; i < DRAW_PARAM_COUNT; ++i)
 	{
 		const auto& p = drawParams[i];
-		luminanceProbeProgram_->setUniformValue(luminanceProbeUniformLocations_.colorScale,
-		                                        p.relativeBrightness * p.eclipseFactor * p.fboColorScale);
+		luminanceProbeProgram_->setUniformValue(luminanceProbeUniformLocations_.isMoon,
+		                                        i == DRAW_PARAM_MOON);
 		luminanceProbeProgram_->setUniformValue(luminanceProbeUniformLocations_.sunDir, p.dir);
 		GL(gl.glBindTexture(GL_TEXTURE_2D, p.fbo->texture()));
 		GL(gl.glDrawArrays(GL_TRIANGLE_STRIP, 0, 4));
@@ -515,7 +515,13 @@ float AtmosphereLightweight::computeAverageLuminance(const StelCore* core)
 
 	const auto texAvg = textureAverager_->getTextureAverageSimple(luminanceProbeFBO_->texture(),
 	                                                              LUM_PROBE_FBO_WIDTH, LUM_PROBE_FBO_HEIGHT);
-	return texAvg[2];
+	float lum = 0;
+	for (int i = 0; i < DRAW_PARAM_COUNT; ++i)
+	{
+		const auto& p = drawParams[i];
+		lum += texAvg[i] * p.relativeBrightness * p.eclipseFactor * p.fboColorScale;
+	}
+	return lum;
 }
 
 void AtmosphereLightweight::computeDrawParams(const StelCore* core, const Planet* lightSource, const Planet* moon,

--- a/src/core/modules/AtmosphereLightweight.hpp
+++ b/src/core/modules/AtmosphereLightweight.hpp
@@ -54,6 +54,7 @@ private:
 	std::unique_ptr<class QOpenGLFramebufferObject> luminanceProbeFBO_;
 	std::unique_ptr<TextureAverageComputer> textureAverager_;
 	std::vector<float> layerSolarElevations_;
+	std::vector<float> layerValueMaxima_;
 	std::vector<float> layerNorms_;
 	std::vector<unsigned> layerVertexOffsetsInVBO_;
 	std::vector<unsigned> layerIndexOffsetsInBuffer_;
@@ -105,6 +106,11 @@ private:
 		int atmoTex;
 		int projectionMatrixInverse;
 	} luminanceProbeUniformLocations_;
+
+	struct
+	{
+		int colorScale;
+	} prepProgramUniformLocations_;
 
 	void loadMesh();
 	//! Negative are special labels to denote single VAOs, others are offsets in #vaos_

--- a/src/core/modules/AtmosphereLightweight.hpp
+++ b/src/core/modules/AtmosphereLightweight.hpp
@@ -100,9 +100,9 @@ private:
 
 	struct
 	{
+		int isMoon;
 		int sunDir;
 		int atmoTex;
-		int colorScale;
 		int projectionMatrixInverse;
 	} luminanceProbeUniformLocations_;
 

--- a/src/core/modules/AtmosphereShowMySky.cpp
+++ b/src/core/modules/AtmosphereShowMySky.cpp
@@ -281,7 +281,7 @@ void AtmosphereShowMySky::resizeRenderTarget(int width, int height)
 	const int physWidth = width/atmoRes;
 	const int physHeight = height/atmoRes;
 	renderer_->resizeEvent(physWidth, physHeight);
-	textureAverager_.reset(new TextureAverageComputer(*StelOpenGL::highGraphicsFunctions(),
+	textureAverager_.reset(new TextureAverageComputer(StelOpenGL::highGraphicsFunctions(),
 	                                                  physWidth, physHeight, GL_RGBA32F, true));
 
 	prevWidth_=width;

--- a/src/core/modules/AtmosphereShowMySky.cpp
+++ b/src/core/modules/AtmosphereShowMySky.cpp
@@ -282,7 +282,7 @@ void AtmosphereShowMySky::resizeRenderTarget(int width, int height)
 	const int physHeight = height/atmoRes;
 	renderer_->resizeEvent(physWidth, physHeight);
 	textureAverager_.reset(new TextureAverageComputer(*StelOpenGL::highGraphicsFunctions(),
-	                                                  physWidth, physHeight, GL_RGBA32F));
+	                                                  physWidth, physHeight, GL_RGBA32F, true));
 
 	prevWidth_=width;
 	prevHeight_=height;

